core/src/main/scala/akka/persistence/cassandra/journal/TagWriter.scala

```diff
@@ -0,0 +1,303 @@
+    case TagWrites(payload) =>
+      validate(payload)
+      // buffer until current query is finished
+      become(writeInProgress(buffer ++ payload))
+      sender() ! TagWriteAck
+    case TagWriteDone(summary) =>
+      summary.foreach {
+        case (id, p @ PidProgress(_, seqNrTo, tagPidSequenceNr, offset)) =>
+          log.debug("Writing tag progress {}", p)
+          // These writes do not block future writes. We don't read he tag progress again from C*
+          // until a restart has happened. This is best effort and expect recovery (not implemented yet
+          // to fix up any missing)
+          // TODO Deal with failure and add debug logging for when done
+          session.writeProgress(id, seqNrTo, tagPidSequenceNr, offset)
+      }
+      log.debug(s"Tag write complete. ${summary}")
+      flushIfRequired(buffer)
+      pubsub.foreach {
+        log.debug("Publishing tag update for {}", tag)
+        _.mediator ! DistributedPubSubMediator.Publish("akka.persistence.cassandra.journal.tag", tag)
+      }
+  }
+
+  private def flushIfRequired(buffer: Vector[Serialized]): Unit = {
+    if (buffer.isEmpty) {
+      context.become(idle(buffer))
+    } else if (buffer.head.timeBucket < buffer.last.timeBucket) {
+      if (log.isDebugEnabled) {
+        log.debug("Switching time buckets: head: {} last: {}", buffer.head.timeBucket, buffer.last.timeBucket)
+      }
+      val (currentBucket, rest) = buffer.span(_.timeBucket == buffer.head.timeBucket)
+      write(currentBucket, rest)
+    } else if (buffer.size >= settings.maxBatchSize) {
+      log.debug("Batch size reached. Writing msg")
+      write(buffer.take(settings.maxBatchSize), buffer.drop(settings.maxBatchSize))
+    } else if (settings.flushInterval == Duration.Zero) {
+      log.debug("Flushing right away as interval is zero")
+      write(buffer, Vector.empty[Serialized])
+    } else {
+      timers.startSingleTimer(FlushKey, Flush, settings.flushInterval)
+      log.debug("Batch size not reached, buffering. Current buffer size: {}", buffer.size)
+      context.become(idle(buffer))
+    }
+  }
+
+  /**
+   * The first time we see a pid we go to Cassandra to see if this if this
+   * tag/pid has been active before.
+   * No other ActorSystem should be reading/updating this row as a pid is owned
+   * by one node.
+   * If the pid/tag combo has no previous progress we start from 1 otherwise
+   * the old pidTagSequenceNr + 1.
+   * The future for getting the progress from C* is added to the lookup map
+   * so any further requests for that pid have to wait until the initial read
+   * is complete.
+   */
+  private def calculateTagPidSequenceNr(pid: String, bucket: TimeBucket): Future[Long] = {
+    // Have we got it already?
+    val tagPidSequenceNr: Future[SequenceNr] = pidTagSequenceNrs.get(pid) match {
+      case None =>
+        log.debug("First time seeing this pId {} since startup", pid)
+        session.selectTagProgress(pid).flatMap {
```

Comment:
what happens if this query fails? the failed future will be placed in the Map and will never recover?

Reply:
I haven't dealt with Cassandra failures yet, will address those